pgml-cms/blog/semantic-search-in-postgres-in-15-minutes.md

```diff
@@ -0,0 +1,467 @@
+---
+description: >-
+  Learn how to implement semantic search in PostgreSQL with nothing but SQL.
+featured: true
+tags: ["Engineering"]
+---
+
+# Semantic Search in Postgres in 15 Minutes
+
+<div align="left">
+
+<figure><img src=".gitbook/assets/silas.jpg" alt="Author" width="125"><figcaption></figcaption></figure>
+
+</div>
+
+Silas Marvin
+
+June 15, 2024
+
+## What is and is not semantic search
+
+Semantic search uses machine learning to understand the meaning of text by converting it into numerical vectors, allowing for more accurate and context-aware search results.
+
+When users are unsure of the exact terms to search for, semantic search can uncover relevant information that traditional keyword searches might miss. This capability is particularly valuable for discovering content based on the intent and context of the search query, rather than relying solely on precise word matches.
+
+It is not a complete replacement for full-text search. In many cases, full-text search can outperform semantic search. Specifically, if a user knows the exact phrase they want to match in a document, full-text search is faster and guaranteed to return the correct result, whereas semantic search is only likely to return the correct result. Full-text search and semantic search can be combined to create powerful and robust search systems.
+
+Semantic search is not just for machine learning engineers. The system behind semantic search is relatively easy to implement, and thanks to new Postgres extensions like _pgml_ and _pgvector_, it is readily available to SQL developers. Just as modern SQL developers are expected to be familiar with and capable of implementing full-text search, they will soon be expected to implement semantic search as well.
+
+## Embeddings 101
+
+Semantic search is powered by embeddings. To understand how semantic search works, we must have a basic understanding of embeddings. 
+
+Embeddings are vectors. Given some text and some embedding model, we can convert text to vectors:
```

Comment:
more programmers are familiar with arrays than vectors if you're looking for a common explanation.